File: web/src/utils/units.ts
const KM_TO_MI = 0.621371
const M_TO_MI = 0.000621371
const KM_TO_YD = 1093.6132983377
const M_TO_YD = 1.0936132983377

export function isSwimDiscipline(value: string | null | undefined): boolean {
  if (!value) return false
  const normalized = value.trim().toLowerCase()
  return normalized.includes('swim') || normalized.includes('pool') || normalized.includes('wetsuit')
}

export function kilometersToMiles(km: number): number {
  return km * KM_TO_MI
}

export function kilometersToYards(km: number): number {
  return km * KM_TO_YD
}

export function metersToMiles(meters: number): number {
  return meters * M_TO_MI
}

export function metersToYards(meters: number): number {
  return meters * M_TO_YD
}

export function formatDistanceFromMeters(
  meters: number | null | undefined,
  discipline?: string | null,
): string {
  if (meters == null || meters <= 0) return '-'
  if (isSwimDiscipline(discipline)) {
    return `${Math.round(metersToYards(meters)).toLocaleString()} yd`
  }
  return `${metersToMiles(meters).toFixed(1)} mi`
}

export function formatDistanceFromKilometers(
  kilometers: number | null | undefined,
  discipline?: string | null,
): string {
  if (kilometers == null || kilometers <= 0) return '-'
  if (isSwimDiscipline(discipline)) {
    return `${Math.round(kilometersToYards(kilometers)).toLocaleString()} yd`
  }
  return `${kilometersToMiles(kilometers).toFixed(1)} mi`
}

File: api/src/services/units.py
"""Shared distance and pace formatting helpers for API responses."""

KM_TO_MI = 0.621371192237334
KM_TO_YD = 1093.6132983377079
M_TO_MI = 0.0006213711922373339
M_TO_YD = 1.0936132983377078


def is_swim_discipline(value: str | None) -> bool:
    if not value:
        return False
    normalized = value.strip().lower()
    return (
        "swim" in normalized
        or "pool" in normalized
        or "wetsuit" in normalized
    )


def format_distance_from_meters(
    meters: float | None, discipline: str | None = None
) -> str:
    if meters is None or meters <= 0:
        return "-"
    if is_swim_discipline(discipline):
        return f"{round(meters * M_TO_YD):,} yd"
    return f"{meters * M_TO_MI:.1f} mi"


def format_distance_from_kilometers(
    kilometers: float | None, discipline: str | None = None
) -> str:
    if kilometers is None or kilometers <= 0:
        return "-"
    if is_swim_discipline(discipline):
        return f"{round(kilometers * KM_TO_YD):,} yd"
    return f"{kilometers * KM_TO_MI:.1f} mi"


def format_pace_per_mile(seconds_per_km: float | None) -> str:
    if seconds_per_km is None or seconds_per_km <= 0:
        return "-"
    seconds_per_mile = int(round(seconds_per_km * 1.60934))
    minutes, seconds = divmod(seconds_per_mile, 60)
    return f"{minutes}:{seconds:02d}/mi"
